core/src/main/java/com/gentics/mesh/plugin/registry/DelegatingPluginRegistryImpl.java

```diff
@@ -0,0 +1,204 @@
+		}
+	}
+
+	@Override
+	public Completable deregister(MeshPlugin plugin) {
+		Objects.requireNonNull(plugin, "The plugin must not be null");
+		String id = plugin.id();
+		log.debug("Deregistering plugin {}", id);
+		return registries().flatMapCompletable(r -> r.deregister(plugin));
+	}
+
+	@Override
+	public synchronized void checkForConflict(MeshPlugin plugin) {
+		for (PluginRegistry registry : Arrays.asList(graphqlRegistry, restRegistry)) {
+			registry.checkForConflict(plugin);
+		}
+	}
+
+	@Override
+	public void preRegister(MeshPlugin plugin) {
+		EventBus eb = rxVertx.get().eventBus();
+		Objects.requireNonNull(plugin, "The plugin must not be null");
+		manager.get().setStatus(plugin.id(), PluginStatus.PRE_REGISTERED);
+		// TODO add payload
+		eb.publish(MeshEvent.PLUGIN_PRE_REGISTERED.getAddress(), null);
+		initAndRegister(plugin);
+	}
+
+	private Observable<PluginRegistry> registries() {
+		return Observable.fromArray(graphqlRegistry, restRegistry, authServiceRegistry);
+	}
+
+	/**
+	 * Register all currently pre registered plugins.
+	 */
+	private void initAndRegister(MeshPlugin plugin) {
+		EventBus eb = rxVertx.get().eventBus();
+		if (log.isDebugEnabled()) {
+			log.debug("Invoking registration of pre-registered plugins");
+		}
+		long timeout = getPluginTimeout().getSeconds();
+
+		String id = plugin.id();
+
+		optionalLock(registerAndInitalizePlugin(plugin)).subscribe(() -> {
+			log.info("Completed handling of pre-registered plugin {" + id + "}");
+			manager.get().setStatus(id, REGISTERED);
+			eb.publish(MeshEvent.PLUGIN_REGISTERED.getAddress(), null);
+			eb.publish(MeshEvent.PLUGIN_DEPLOYED.getAddress(), null);
+		}, err -> {
+			if (err instanceof TimeoutException) {
+				log.error("The registration of plugin {" + id + "} did not complete within {" + timeout
+					+ "} seconds. Unloading plugin.");
+			} else {
+				log.error("Plugin init and register failed for plugin {" + id + "}", err);
+			}
+			manager.get().setStatus(id, FAILED);
+			eb.publish(MeshEvent.PLUGIN_DEPLOY_FAILED.getAddress(), null);
+		});
+
+	}
+
+	private Completable registerAndInitalizePlugin(MeshPlugin plugin) {
+		long timeout = getPluginTimeout().getSeconds();
+		String id = plugin.id();
+
+		return optionalQuorumCheck().doOnComplete(() -> {
+			if (log.isDebugEnabled()) {
+				log.debug("Invoking initialization of plugin {" + id + "}");
+			}
+		}).andThen(plugin.initialize().timeout(timeout, TimeUnit.SECONDS).doOnComplete(() -> {
+			manager.get().setStatus(id, INITIALIZED);
+		}).andThen(register(plugin).doOnComplete(() -> {
+			manager.get().setStatus(id, REGISTERED);
+		})));
+	}
+
+	/**
+	 * Use the lock in clustered mode to prevent concurrent inits
+	 * 
+	 * @param lockedAction
+	 * @return
+	 */
+	private Completable optionalLock(Completable lockedAction) {
+		if (options.getClusterOptions().isEnabled()) {
+			return rxVertx.get().sharedData().rxGetLockWithTimeout(GLOBAL_PLUGIN_LOCK_KEY, 10_000).toMaybe()
```

Comment:
This should ideally be configurable or at least be derived from the plugin initialize timeout